docs/spec/relayer/Relayer.md

```diff
@@ -220,20 +71,123 @@ More details about light client assumptions and protocols can be found
-`verifyMembership(header.AppHash, h, proof, path, d)` that returns `true` if data was committed by the corresponding
+`VerifyMembership(header.AppHash, h, proof, path, d)` that returns `true` if data was committed by the corresponding
 chain at height *h*. The trusted header is provided by the corresponding light client. 
   
+## Relayer algorithm
+
+The main relayer event loop is a pipeline of four stages. Assuming some IBC event at height `h` on `chainA`, 
+the relayer:
+
+1. Determine destination chain (`chainB`)
+2. Updates (on `chainB`) the IBC client for `chainA` to a certain height `H` where `H >= h+1`.
+3. Create IBC datagram at height `H-1`.
+4. Submit the datagram from stage (2) to `chainB`.
+
+Note that an IBC event at height `h` corresponds to the modifications to the data store made as part of executing
+block at height `h`. The corresponding proof (that data is indeed written to the data store) can be verified using
+the data store root hash that is part of the header at height `h+1`.
+
+Once stage 2 finishes correctly, stage 3 should succeed assuming that `chainB` has not already processed the event. The 
+interface between stage 2 and stage 3 is just the height `H`. Once stage 3 finishes correctly, stage 4 should 
+succeed. The interface between stage 3 and stage 4 is an IBC datagram.
+
+We assume that the corresponding light client is correctly installed on each chain.
+
+The semantic of the returned error is to signal to the caller whether there was an error in the event processing that 
+could be transient (for example not able to establish connection to a correct full node), in which case
+the event can be reprocessed later (`Error.RETRY`), or the event should be dropped as it has already been 
+received by the destination chain due to activity of a concurrent relayer (`Error.DROP`).
+
+Data structures and helper function definitions are provided 
+[here](https://github.com/informalsystems/ibc-rs/blob/master/docs/spec/relayer/Definitions.md). 
+
+```golang
+enum Error {
+  RETRY,
+  DROP,
+}
+```
+
+```golang
+func handleEvent(ev, chainA) Error {
+    // NOTE: we don't verify if event data are valid at this point. We trust full node we are connected to
+    // until some verification fails. 
+    
+    // Stage 1.
+    // Determine destination chain
+    chainB = getDestinationInfo(ev, chainA) 
+    if chainB == nil { return Error.RETRY }  
+
+    // Stage 2.
+    // Update on `chainB` the IBC client for `chainA` to height `>= targetHeight`.
+    targetHeight = ev.height + 1
+    // See the code for `updateIBCClient` below.
+    installedHeight, error := updateIBCClient(chainB, chainA, targetHeight)
+    if error != nil { return Error.RETRY }
+
+    // Stage 3.
+    // Create the IBC datagrams including `ev` & verify them.
+    datagram, error = CreateDatagram(ev, chainA, chainB, installedHeight)
+    if error != nil { return Error.RETRY }
```

Comment:
Need to tell the caller if it should drop or retry? So just pass the error through
```suggestion
    if error != nil { return error }
```